File: Pickleball Performance Dashboard/src/lib/analysis/contract.ts
// TypeScript mirror of server/picklepro/contract.py (schema version 1.0).
// contract.test.ts checks these lists against contracts/analysis_result.v1.schema.json
// so the two sides cannot drift silently.

export const RESULT_STATUSES = ["ok", "insufficient_data"] as const;
export const DATA_ORIGINS = ["measured", "test_fixture"] as const;
export const METRIC_STATUSES = ["measured", "insufficient_data", "not_computed", "experimental"] as const;
export const VALIDATION_LEVELS = ["not_evaluated", "synthetic_only", "evaluated_on_real_footage"] as const;
export const METRIC_KEYS = ["court_heatmap", "zone_occupancy", "rally_segmentation", "shot_classification"] as const;
export const RESULT_KEYS = [
  "schema_version", "status", "data_origin", "message", "provenance", "video", "coverage", "calibration",
  "player_selection", "tracks", "player_positions", "metrics", "warnings",
] as const;

export type ResultStatus = (typeof RESULT_STATUSES)[number];
export type DataOrigin = (typeof DATA_ORIGINS)[number];
export type MetricStatus = (typeof METRIC_STATUSES)[number];
export type ValidationLevel = (typeof VALIDATION_LEVELS)[number];
export type MetricKey = (typeof METRIC_KEYS)[number];

export type Metric = {
  status: MetricStatus;
  validation: ValidationLevel;
  scope: "whole_clip";
  reason: string | null;
};

export type HeatmapValue = {
  units: "seconds";
  coordinate_system: string;
  cell_size_m: number;
  x_edges_m: number[];
  y_edges_m: number[];
  dwell_seconds: number[][];
  tracked_time_s: number;
  outside_mapped_area_s: number;
};

export type ZoneOccupancyValue = {
  zone_definitions: Record<string, string>;
  seconds: Record<string, number>;
  fraction_of_tracked_time: Record<string, number>;
};

export type PlayerBox = { track_id: number | null; bbox: [number, number, number, number]; confidence: number | null };
export type PositionSnapshot = { time_seconds: number; players: PlayerBox[] };

export type AnalysisResultV1 = {
  schema_version: "1.0";
  status: ResultStatus;
  data_origin: DataOrigin;
  message: string;
  provenance: {
    pipeline_version: string;
    generated_at: string;
    detector: { name: string; confidence_is_model_score: boolean };
    source: { filename: string | null; sha256: string | null };
  };
  video: {
    width: number;
    height: number;
    fps: number;
    frame_count_reported: number | null;
    container_duration_s: number | null;
  };
  coverage: {
    analyzed_start_s: number;
    analyzed_end_s: number;
    analyzed_duration_s: number;
    frames_decoded: number;
    frames_analyzed: number;
    sample_stride: number;
    decode_failures: number;
    stopped_early_reason: string | null;
    fraction_of_video_analyzed: number | null;
    frames_with_detections: number;
  };
  calibration: {
    method: "manual_landmarks" | "auto_model_landmarks";
    court_model: string;
    landmarks_used: string[];
    reprojection_rmse_px: number;
    reprojection_rmse_m: number;
    quality: "good" | "poor";
  } | null;
  player_selection: {
    method: "track_id" | "court_half";
    track_id: number | null;
    court_half: "near" | "far" | null;
    tracked_time_s: number;
    tracked_fraction: number;
    ambiguous_frames: number;
  } | null;
  tracks: { track_id: number; first_seen_s: number; last_seen_s: number; observed_frames: number }[];
  player_positions: PositionSnapshot[];
  metrics: {
    court_heatmap: Metric & { value: HeatmapValue | null };
    zone_occupancy: Metric & { value: ZoneOccupancyValue | null };
    rally_segmentation: Metric;
    shot_classification: Metric;
  };
  warnings: string[];
};

export class ContractError extends Error {}

const isObj = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);

function oneOf<T extends string>(v: unknown, allowed: readonly T[], path: string): T {
  if (typeof v !== "string" || !(allowed as readonly string[]).includes(v)) {
    throw new ContractError(`${path}: expected one of ${allowed.join(", ")}, got ${JSON.stringify(v)}`);
  }
  return v as T;
}

function num(v: unknown, path: string): number {
  if (typeof v !== "number" || !Number.isFinite(v)) throw new ContractError(`${path}: expected a number`);
  return v;
}

/**
 * Validate the parts of a result the UI relies on to describe it truthfully.
 * Unknown or malformed results are rejected rather than rendered as if valid.
 */
export function parseAnalysisResult(input: unknown): AnalysisResultV1 {
  if (!isObj(input)) throw new ContractError("result: expected an object");
  if (input.schema_version !== "1.0") {
    throw new ContractError(`schema_version: unsupported ${JSON.stringify(input.schema_version)}`);
  }
  oneOf(input.status, RESULT_STATUSES, "status");
  oneOf(input.data_origin, DATA_ORIGINS, "data_origin");
  if (typeof input.message !== "string") throw new ContractError("message: expected a string");
  const prov = input.provenance;
  if (!isObj(prov) || typeof prov.pipeline_version !== "string" || !isObj(prov.detector)) {
    throw new ContractError("provenance: missing pipeline_version or detector");
  }
  const cov = input.coverage;
  if (!isObj(cov)) throw new ContractError("coverage: expected an object");
  num(cov.analyzed_duration_s, "coverage.analyzed_duration_s");
  num(cov.frames_analyzed, "coverage.frames_analyzed");
  if (input.calibration !== null) {
    if (!isObj(input.calibration)) throw new ContractError("calibration: expected an object or null");
    oneOf(input.calibration.method, ["manual_landmarks", "auto_model_landmarks"] as const, "calibration.method");
  }
  if (!isObj(input.metrics)) throw new ContractError("metrics: expected an object");
  for (const key of METRIC_KEYS) {
    const m = (input.metrics as Record<string, unknown>)[key];
    if (!isObj(m)) throw new ContractError(`metrics.${key}: missing`);
    oneOf(m.status, METRIC_STATUSES, `metrics.${key}.status`);
    oneOf(m.validation, VALIDATION_LEVELS, `metrics.${key}.validation`);
  }
  const heat = (input.metrics as Record<string, Record<string, unknown>>).court_heatmap;
  if (heat.status === "measured" && !isObj(heat.value)) {
    throw new ContractError("metrics.court_heatmap: measured without a value");
  }
  if (!Array.isArray(input.player_positions) || !Array.isArray(input.tracks) || !Array.isArray(input.warnings)) {
    throw new ContractError("player_positions/tracks/warnings: expected arrays");
  }
  return input as unknown as AnalysisResultV1;
}

export const METRIC_LABELS: Record<MetricKey, string> = {
  court_heatmap: "Court heatmap (dwell time)",
  zone_occupancy: "Zone occupancy",
  rally_segmentation: "Rally segmentation",
  shot_classification: "Shot classification",
};

export const VALIDATION_LABELS: Record<ValidationLevel, string> = {
  not_evaluated: "Accuracy not yet evaluated",
  synthetic_only: "Checked on synthetic video only",
  evaluated_on_real_footage: "Evaluated on real footage",
};

File: Pickleball Performance Dashboard/server/picklepro/contract.py
"""Analysis result contract (schema version 1).

This module is the single source of truth for the result shape. The TypeScript
mirror lives in ``src/lib/analysis/contract.ts`` and both are checked against
``contracts/analysis_result.v1.schema.json`` by the test suites.

Design rules:

* A job that could not run is a *job failure*, not a result. A result only
  exists when the video was read.
* ``status == "insufficient_data"`` is a legitimate outcome and must never be
  rewritten into a success message.
* Every metric carries its own status and validation level so the UI can say
  exactly what is measured, what is missing, and what has not been validated.
* ``confidence`` is ``None`` unless a model actually produced a score.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

SCHEMA_VERSION = "1.0"

ResultStatus = Literal["ok", "insufficient_data"]
DataOrigin = Literal["measured", "test_fixture"]
MetricStatus = Literal["measured", "insufficient_data", "not_computed", "experimental"]
# How far a metric's accuracy has been checked. Nothing is "validated" until it
# has been compared against labelled real footage.
ValidationLevel = Literal["not_evaluated", "synthetic_only", "evaluated_on_real_footage"]
MetricScope = Literal["whole_clip"]


class _Model(BaseModel):
    model_config = ConfigDict(extra="forbid")


class DetectorInfo(_Model):
    name: str
    confidence_is_model_score: bool = Field(
        description="True only when per-detection confidence comes from a model. "
        "Motion-based detection has no confidence and reports null."
    )


class SourceInfo(_Model):
    filename: Optional[str] = None
    sha256: Optional[str] = None


class Provenance(_Model):
    pipeline_version: str
    generated_at: str
    detector: DetectorInfo
    source: SourceInfo


class VideoInfo(_Model):
    width: int
    height: int
    fps: float
    frame_count_reported: Optional[int] = Field(
        None, description="Frame count from the container header; may be approximate."
    )
    container_duration_s: Optional[float] = Field(
        None, description="frame_count_reported / fps; null when the container does not report it."
    )


class Coverage(_Model):
    analyzed_start_s: float
    analyzed_end_s: float
    analyzed_duration_s: float
    frames_decoded: int
    frames_analyzed: int
    sample_stride: int = Field(description="Every Nth decoded frame was analyzed.")
    decode_failures: int
    stopped_early_reason: Optional[str] = Field(
        None, description="Set when analysis stopped before the end of the video (e.g. max_seconds)."
    )
    fraction_of_video_analyzed: Optional[float] = None
    frames_with_detections: int


class CalibrationSummary(_Model):
    method: Literal["manual_landmarks", "auto_model_landmarks"]
    court_model: str
    landmarks_used: List[str]
    reprojection_rmse_px: float
    reprojection_rmse_m: float
    quality: Literal["good", "poor"]


class PlayerSelectionSummary(_Model):
    method: Literal["track_id", "court_half"]
    track_id: Optional[int] = None
    court_half: Optional[Literal["near", "far"]] = None
    tracked_time_s: float
    tracked_fraction: float = Field(description="tracked_time_s / coverage.analyzed_duration_s")
    ambiguous_frames: int = Field(0, description="Frames excluded because the selection matched >1 detection.")


class TrackSummary(_Model):
    track_id: int
    first_seen_s: float
    last_seen_s: float
    observed_frames: int


class PlayerBox(_Model):
    track_id: Optional[int] = None
    bbox: List[int] = Field(min_length=4, max_length=4)
    confidence: Optional[float] = None


class PositionSnapshot(_Model):
    time_seconds: float
    players: List[PlayerBox]


class Metric(_Model):
    status: MetricStatus
    validation: ValidationLevel = "not_evaluated"
    scope: MetricScope = "whole_clip"
    reason: Optional[str] = Field(None, description="Why the metric is missing, partial, or experimental.")


class HeatmapValue(_Model):
    units: Literal["seconds"] = "seconds"
    coordinate_system: str
    cell_size_m: float
    x_edges_m: List[float]
    y_edges_m: List[float]
    dwell_seconds: List[List[float]] = Field(description="rows follow y (near→far), columns follow x (left→right)")
    tracked_time_s: float
    outside_mapped_area_s: float


class CourtHeatmapMetric(Metric):
    value: Optional[HeatmapValue] = None


class ZoneOccupancyValue(_Model):
    zone_definitions: Dict[str, str]
    seconds: Dict[str, float]
    fraction_of_tracked_time: Dict[str, float]


class ZoneOccupancyMetric(Metric):
    value: Optional[ZoneOccupancyValue] = None


class Metrics(_Model):
    court_heatmap: CourtHeatmapMetric
    zone_occupancy: ZoneOccupancyMetric
    rally_segmentation: Metric
    shot_classification: Metric


class AnalysisResultV1(_Model):
    schema_version: Literal["1.0"] = SCHEMA_VERSION
    status: ResultStatus
    data_origin: DataOrigin
    message: str
    provenance: Provenance
    video: VideoInfo
    coverage: Coverage
    calibration: Optional[CalibrationSummary] = None
    player_selection: Optional[PlayerSelectionSummary] = None
    tracks: List[TrackSummary]
    player_positions: List[PositionSnapshot]
    metrics: Metrics
    warnings: List[str]


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def not_computed(reason: str) -> Metric:
    return Metric(status="not_computed", reason=reason)


# Reasons for metrics this pipeline deliberately does not produce yet.
RALLY_NOT_COMPUTED = (
    "Rally segmentation is not implemented or validated. All metrics are whole-clip metrics."
)
SHOTS_NOT_COMPUTED = (
    "Shot classification requires ball tracking and a validated classifier; neither exists in this pipeline yet."
)


def json_schema() -> dict:
    return AnalysisResultV1.model_json_schema()
